### PR TITLE
Zookeeper backed leasing

### PR DESCRIPTION
**Pull request by mmihic**

### CODE REVIEW

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/AtomicEnum.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package misk.clustering.zookeeper
+
+import java.util.concurrent.atomic.AtomicInteger
+
+/** An [AtomicEnum] is an enum that can be updated atomically, including down an atomic compare and set */
```
Comment (wesleyk) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/AtomicEnum.kt]: not quite following the second part of the sentence: "including down an atomic compare and set"

Comment (wesleyk) [reply misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/AtomicEnum.kt]: (maybe rm down?)

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  fun connectionLost() {
+    lock.withLock {
+      // We're disconnected from the cluster, so we don't know the state of the lease. Zk will
+      // automatically expire the lease if we remain disconnected for too long, and in the
+      // meanwhile we should act as though the lease has been lost
+      if (status != Status.CLOSED) status = Status.UNKNOWN
+    }
+  }
+
+  private fun release(guaranteeDelete: Boolean = true): Boolean {
```
Comment (wesleyk) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt]: is the guaranteeDelete flag ever used as false?

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+ * substantial full JVM pause, the lease may expire if zookeeper doesn't hear from client before
+ * the zookeeper session timeout passes. This means that under certain circumstances, the
+ * leadership durability is tied to the zookeeper session timeout.
+ *
+ * <p>
+ * Users of this lease framework should avoid creating tasks that have a duration longer than
+ * the session timeout or they risk having a split brain running two leader tasks on different hosts
+ * simultaneously.
+ */
+internal class ZkLease(
```
Comment (wesleyk) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt]: code was nice and straight forward to follow 👏

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/AtomicEnum.kt (hunk 1 above)
Comment (mmihic) [reply misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/AtomicEnum.kt]: oops, fixed

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt (hunk 2 above)
Comment (mmihic) [reply misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt]: oh eh only in a previous iteration. removing

File: misk-zookeeper/src/test/kotlin/misk/clustering/zookeeper/ZkLeaseTest.kt [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    assertThat(connected.await(5, TimeUnit.SECONDS)).isTrue()
+  }
+
+  @Test fun acquiresLeaseIfMappedToSelf() {
+    cluster.resourceMapper.addMapping(leasePath, self)
+
+    val lease = leaseManager.requestLease(LEASE_NAME)
+    assertThat(lease.checkHeld()).isTrue()
+    assertThat(lease.checkHeld()).isTrue()
```
Comment (wesleyk) [inline misk-zookeeper/src/test/kotlin/misk/clustering/zookeeper/ZkLeaseTest.kt]: rm redundant assertion

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  private fun currentLeaseData() = try {
+    manager.client.value.data.forPath(leaseZkPath)
+  } catch (e: KeeperException.NoNodeException) {
+    // The lease node no longer exists
+    null
+  }
+
+  /** Called when we detect that the lease is being actively held by another instance */
+  private fun leaseHeldByAnother() {
+    synchronized(this) {
```
Comment (ryanhall07) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt]: why are you not using the lock?

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+          .withMode(CreateMode.EPHEMERAL)
+          .withACL(ZooDefs.Ids.OPEN_ACL_UNSAFE)
+          .forPath(leaseZkPath, leaseData)
+      status = Status.HELD
+      log.info { "acquired lease $name" }
+      return true
+    } catch (e: KeeperException.NodeExistsException) {
+      leaseHeldByAnother()
+    } catch (e: Exception) {
+      log.info(e) { "got unexpected exception trying to acquire node for lease $name" }
```
Comment (ryanhall07) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt]: I think this should be a warn if it's unexpected.

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+      return false
+    } catch (e: Error) {
+      log.error(e) { "unexpected error checking if lease $name is held" }
+      throw e
+    }
+  }
+
+  fun close() {
+    lock.withLock {
+      if (status == Status.HELD) release()
```
Comment (ryanhall07) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt]: should we just always call `release`? what if the status is `Unknown`? probably still worth trying a best effort release

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLeaseManager.kt [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    connected.set(false)
+
+    // When a cluster change occurred, inform all of the leases to recheck their ownership
+    cluster.watch {
+      actionQueue.put {
+        handleClusterChange()
+      }
+    }
+
+    // When we disconnect from zk, inform all of the leases that their state is no unknown
```
Comment (ryanhall07) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLeaseManager.kt]: i don't follow the comment

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLeaseManager.kt [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    })
+  }
+
+  override fun triggerShutdown() {
+    log.info { "stopping zk lease manager" }
+
+    if (!state.compareAndSet(State.RUNNING, State.STOPPING)) return
+
+    // Clear any pending actions, release any leases we currently hold, and stop running
+    val fullyStopped = CountDownLatch(1)
```
Comment (ryanhall07) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLeaseManager.kt]: nit: might be nice to await the `actionQueue` is cleared

File: misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkService.kt [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    curatorFramework.start()
+  }
+
+  override fun shutDown() {
+    log.info { "closing connection to zookeeper" }
+
+    try {
+      curatorFramework.close()
+    } catch (th: Throwable) {
+      log.error(th) { "unable to close connection to zookeeper" }
```
Comment (ryanhall07) [inline misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkService.kt]: does the `Service` not log exceptions for you?

File: dependencies.gradle [hunk 11]
```diff
@@ -73,5 +74,7 @@ ext.dep = [
   "vitess": "io.vitess:vitess-jdbc:2.2.0",
   "wireRuntime": "com.squareup.wire:wire-runtime:2.3.0-RC1",
   "zipkinBrave": "io.zipkin.brave:brave:4.17.2",
-  "zipkinReporter": "io.zipkin.reporter2:zipkin-sender-okhttp3:2.4.1"
+  "zipkinReporter": "io.zipkin.reporter2:zipkin-sender-okhttp3:2.4.1",
```
[the remark this answers is not in the document]
Comment (mmihic) [reply dependencies.gradle]: I think we used to and maybe still do have a zipkin reporter?

**Pull request merged**